### PR TITLE
Switch to poetry and publish pypi package

### LINKED ISSUE #30: Publish semgrep-agent to pypi
We can rework it to use poetry for this.

### PR DESCRIPTION
**Pull request by brendongo**

### CODE REVIEW

**Comment by brendongo** [comment]
Still need to update development instructions

**Comment by brendongo** [comment]
Because of https://github.com/python-poetry/poetry/issues/2711 poetry does not work when you want to use python 3.8
Because of https://github.com/python-poetry/poetry/issues/2876 we cannot use the suggested fix of using beta release 1.1.0b1

Will make semgrep-agent 3.7 compliant

**Comment by brendongo** [comment]
Note that with this semgrep version will need to be bumped in pyproject.toml when updating the docker image.
This is so that semgrep can be listed as a dependency in the pypi published semgrep-agent. Alternatively we can have users come with their own semgrep installation. Figure it's less headache to have semgrep as a python dependency.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -2,16 +2,22 @@
 
 ## Development setup
 
-Install dependencies with
+Install the correct version of poetry:
 
 ```
-pipenv install
+pip3 install poetry==poetry==1.1.0b2
```
Comment (underyx) [inline CONTRIBUTING.md]: Is this still the correct version?

File: Dockerfile [hunk 2]
```diff
@@ -1,19 +1,24 @@
 FROM returntocorp/semgrep:0.22.0@sha256:f9c9f9712557599b54658d0ccde32b70f36c26caaf5c64591eabad857b205caa AS semgrep
-FROM python:3.8-alpine
+FROM python:3.7-alpine
 
 WORKDIR /app
-COPY Pipfile* ./
+COPY poetry.lock ./
+COPY pyproject.toml ./
+RUN ls
```
Comment (underyx) [inline Dockerfile]:
```suggestion
```
I assume this was there for debugging.

File: Dockerfile [hunk 3]
```diff
@@ -1,19 +1,24 @@
...
 
 RUN apk add --no-cache --virtual=.build-deps build-base libffi-dev openssl-dev &&\
     apk add --no-cache --virtual=.run-deps bash git less libffi openssl &&\
-    pip install --no-cache-dir pipenv==2020.5.28 &&\
-    pipenv install --system &&\
-    PRECOMPILED_LOCATION=/tmp/semgrep-core pipx install semgrep==${INSTALLED_SEMGREP_VERSION} &&\
-    pip uninstall -y pipenv &&\
+    pip install --no-cache-dir poetry==1.0.10 &&\
+    # Need to install semgrep before everything else so we dont try to install from source
+    PRECOMPILED_LOCATION=/tmp/semgrep-core pip install semgrep==${INSTALLED_SEMGREP_VERSION} &&\
```
Comment (underyx) [inline Dockerfile]: Can we keep `pipx`? It was there to keep semgrep isolated and make sure its dependencies can never conflict with the agent's.

File: src/semgrep_agent/meta.py [hunk 4]
```diff
@@ -96,7 +96,8 @@ def glom_event(self, spec: TType) -> Any:
 
     @cachedproperty
     def event(self) -> Dict[str, Any]:
-        if value := os.getenv("GITHUB_EVENT_PATH"):
```
Comment (underyx) [inline src/semgrep_agent/meta.py:99]: Bye walrus operator, was fun knowing ya 😢

**Comment by underyx** [comment]
Note the merge conflict after the absolute imports change 😇

File: CONTRIBUTING.md (hunk 1 above)
Comment (brendongo) [reply CONTRIBUTING.md]: Nope fixing

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -2,16 +2,22 @@
 
 ## Development setup
 
-Install dependencies with
+Install the correct version of poetry:
 
 ```
-pipenv install
+pip3 install poetry==poetry==1.0.10
```
Comment (underyx) [inline CONTRIBUTING.md]:
```suggestion
pip3 install poetry==1.0.10
```

File: pyproject.toml [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+glom = "^20.8.0"
+pymmh3 = "^0.0.5"
+
+[tool.poetry.dev-dependencies]
+
+[tool.poetry.scripts]
+semgrep-agent = "semgrep_agent.__main__:main"
+
+[build-system]
+requires = ["poetry>=1.1"]
```
Comment (underyx) [inline pyproject.toml]: Do we need to update this as well to 1.0?

**Review by underyx — APPROVED** [review]
wheeeeeeee

**Pull request merged**